dialogue-annotations/src/main/java/com/palantir/dialogue/annotations/ErrorHandlingDeserializer.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.annotations;
+
+import com.palantir.dialogue.Deserializer;
+import com.palantir.dialogue.Response;
+import com.palantir.logsafe.Preconditions;
+import java.util.Optional;
+
+public class ErrorHandlingDeserializer<T> implements Deserializer<T> {
```

Comment:
This should be final and based on the `ErrorHandlingDeserializerFactory` `Deserizlizer` implementation instead

Reply:
Since I'm trying to pass the BodySerDe's deserializers, would you recommend making a custom `DeserializerFactory` class that just returns the BodySerDe ones? Should I maybe make DeserializerFactory into a functional interface and pass BodySerDe#deserializer

Reply:
I think we should follow the pattern used by `ResponseDeserializer`, define an InputStreamDeserialzier + OptionalInputStreamDeserialzier which extend StdDeserializer, and use those directly from codegen

Reply:
Let's ignore OptionalInputStreamDeserialzier initially as well, since that will require some more complex reflection on `Optional<InputStream>` to prevent the error handling deserializer from closing the response prematurely. We can handle that in a separate PR

Reply:
I tried an approach - let me know what you think of it :)

Reply:
I like the idea of passing through the BodySerDe (or even full runtime reference), but I'd prefer to investigate that change separately from this one -- I think we want this implementation to match `Json.java` which handles the default json serialization structure: https://github.com/palantir/dialogue/blob/60bfc2aca0cd6b2054cfe1887b62e20c3d01c9c6/dialogue-annotations/src/main/java/com/palantir/dialogue/annotations/Json.java#L43-L50

Reply:
Ah got it! Let me look into that and push something through! Would you prefer I extract the `DEFAULT_BODY_SERDE` as a constant or just create two of them? (I'm going to go for a shared constant unless you think it's a bad idea)